Remove dependency on pipe events in HttpConnection

davidfowl opened the pull request:
- Refactored the HttpConnection to not depend on OnReaderCompleted and OnWriterCompleted. Instead we use ConnectionClosed to detect the FIN that we propagate via ConnectedAborted.
- Fire ConnectionClosed when a FIN is received from both transports.
- Remove pipe completion from Http1Connection and Http1OutputProducer. Instead just return from request processing.
- Flow the CancellaitonToken property in RawStream

davidfowl committed:
Remove dependency on pipe events
- Refactored the HttpConnection to not depend on OnReaderCompleted and OnWriterCompleted. Instead we use ConnectionClosed to detect the FIN that we propagate via ConnectedAborted.
- Fire ConnectionClosed when a FIN is received from both transports.
- Remove pipe completion from Http1Connection and Http1OutputProducer. Instead just return from request processing.
- Flow the CancellaitonToken property in RawStream

gfoidl commented on src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs:
```diff
@@ -134,6 +142,11 @@ private async Task ReadInputAsync(Stream stream)
+            catch (OperationCanceledException ex)
+            {
+                // Propagate the exception if it's ConnectionAbortedException
+                error = ex is ConnectionAbortedException ? ex : null;
```
```suggestion
                error = ex as ConnectionAbortedException;
```

jkotalik commented:
I'd also prefer to have this for https://github.com/aspnet/AspNetCore/pull/11109

jkotalik commented on src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs:
```diff
@@ -160,19 +160,20 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> http
                         }
                     }
 
-                    _context.Transport.Input.OnWriterCompleted(
```
👏

jkotalik reviewed (approved):
The change itself seems good, but I think we should go over this in person before merging. Also to figure out how we can sync this with https://github.com/aspnet/AspNetCore/pull/11109/checks?check_run_id=146487478. I'd rather merge this first before merging my PR as I don't want to have the pipe scheduler be thread pool.

halter73 commented on src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs:
```diff
@@ -115,9 +124,8 @@ private async Task ReadInputAsync(Stream stream)
-
                     var outputBuffer = Input.Writer.GetMemory(_minAllocBufferSize);
-                    var bytesRead = await stream.ReadAsync(outputBuffer);
+                    var bytesRead = await stream.ReadAsync(outputBuffer, _readingTokenSource.Token);
```
But the perf! JK. looking forward to the update.

davidfowl committed:
Fixed tests
- Cancel the transport input on RawStream to yield the pending read. This is much more efficient than passing a cancellation token into everything (and more reliable)
- Fixed the RequestTests to not depend on inline scheduling
- Properly dispose things in the LibuvOutputConsumerTests

halter73 commented on src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs:
```diff
@@ -44,8 +45,16 @@ public async Task RunAsync(Stream stream)
-            await inputTask;
             await outputTask;
+            // Yield the current request
+            _transport.Input.CancelPendingRead();
```
Shouldn't we wait for the connection middleware pipeline before closing the input? I think this breaks half-duplex if that matters.

halter73 commented on src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs:
```diff
@@ -105,14 +107,19 @@ public override Task FlushAsync(CancellationToken cancellationToken)
                 {
+                    if (_throwOnCancelled && result.IsCanceled)
+                    {
+                        throw new OperationCanceledException();
```
We need a flag to know *we're* the once that canceled the read.

davidfowl replied:
Done

halter73 commented on src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs:
```diff
@@ -134,6 +141,11 @@ private async Task ReadInputAsync(Stream stream)
+            catch (OperationCanceledException ex)
+            {
+                // Propagate the exception if it's ConnectionAbortedException
+                error = ex as ConnectionAbortedException;
```
Why not make this an empty `catch (ConnectionAbortedException) { }` block?

halter73 replied:
I misread. You are only storing this if it *is* a ConnectionAbortedException. In that case having a non-empty ConnectionAbortedException catch then an empty OperationCanceledException catch after is still probably better.

davidfowl replied:
This is less code.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -187,6 +173,10 @@ private async Task DoReceive()
                 // If Shutdown() has already bee called, assume that was the reason ProcessReceives() exited.
                 Input.Complete(_shutdownReason ?? error);
+
+                FireConnectionClosed();
```
This makes me wish we called to ConnectionClosed token FinReceived or something like that, because the connection might be only half closed at this point.

davidfowl replied:
You could argue the same about HttpContext.ConnectionAborted, I don't think it makes a difference.

halter73 replied:
You absolutely could, but at least that's HTTP-specific so it makes a bit more sense.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -307,6 +297,26 @@ private async Task ProcessSends()
+        private void FireConnectionClosed()
+        {
+            // Guard against scheduling this multiple times
+            if (_connectionClosed)
```
I see why you added this bool for the libuv transport, but it seems completely unnecessary in the socket transport. FireConnectionClosed can only ever be called once.

davidfowl replied:
Yes that is true. I noticed that when I was copying code.

halter73 commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs:
```diff
@@ -246,6 +239,26 @@ private void OnRead(UvStreamHandle handle, int status)
+        private void FireConnectionClosed()
+        {
+            // Guard against scheduling this multiple times
+            if (_connectionClosed)
```
I see you're taking advantage of the single-threaded nature of libuv's event loop 👍

halter73 commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs:
```diff
@@ -135,21 +137,10 @@ private async Task StartCore()
-                    // Fire the connection closed token and wait for it to complete
-                    var waitForConnectionClosedTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    // Ensure this always fires
+                    FireConnectionClosed();
```
If we're going to continue firing ConnectionClosed for libuv output errors, we should do the same for the socket transport.

davidfowl replied:
What do you mean? Aren't we doing that? It always fires no?

halter73 replied:
I'm now thinking about making this more like the socket transport instead. The important think is consistency.

Instead of calling FireConnectionClosed here, just don't dispose _socket here. Instead, close the uv_stream_handle only after both the output loop completed and

a) you've received a terminal read callback or
b) the connection was aborted

halter73 commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs:
```diff
@@ -856,6 +852,7 @@ await using (var server = new TestServer(TestApp.EchoAppChunked, testContext))
                     await connection.Send(
                         "POST / HTTP/1.1");
                     connection.ShutdownSend();
+                    await connection.TransportConnection.WaitForCloseTask;
```
Why is this necessary?

halter73 replied:
Discussed this offline. Calling ReceiveEnd unblocks the ProcessRequest loop allowing to observe the FIN and write a 4XX before the transport is aborted.

halter73 commented on src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs:
```diff
@@ -61,17 +63,17 @@ public override int Read(byte[] buffer, int offset, int count)
         public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
         {
-            return ReadAsyncInternal(destination);
+            return ReadAsyncInternal(destination, cancellationToken);
```
Now that we're flowing the cancellationToken through RawStream, do we have a regression test?

halter73 commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs:
```diff
@@ -234,6 +225,8 @@ private void OnRead(UvStreamHandle handle, int status)
                 {
                     Log.ConnectionReadFin(ConnectionId);
+
+                    FireConnectionClosed();
```
Why not fire this for any terminal status?

halter73 replied:
Really we should be calling this every time we call Input.Complete().

davidfowl committed:
PR feedback
- Add CancelPendingRead to RawStream
- Clean up adapted pipeline complete
- Skipped flaky test

davidfowl commented:
@aspnet/build What's the deal with all the PRs, do we merge, do we wait? I know it's being looked at but what is the mitigation to unblock work?

jkotalik commented on src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs:
```diff
@@ -14,11 +14,20 @@ internal sealed class RawStream : Stream
+
+        public void CancelPendingRead()
+        {
+            _cancelCalled = true;
```
Does CancelPendingRead need to be synchronized?

davidfowl replied:
Not for our usage.

halter73 replied:
I don't even think _cancelCalled *needs* to be volatile since CancelPendingRead acquires and releases the Pipe's `_sync` lock right afterwards.

dougbu commented:
@davidfowl as a few people have done, I suggest making a polite request for an admin to merge PRs where build success isn't being reported correctly or individual jobs hang. Looks like both of those issues have hit this PR but is it ready to merge?

davidfowl commented:
> @davidfowl as a few people have done, I suggest making a polite request for an admin to merge PRs where build success isn't being reported correctly or individual jobs hang. Looks like both of those issues have hit this PR but is it ready to merge?

I can merge I just want to know the protocol. Yes this is ready to merge.

dougbu commented:
:ok:, ping if you're blocked
